---
title: "intron_annotation"
author: "Asta_MR"
date: "5/5/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse) 
library(GenomicRanges)
library(GenomicFeatures)
library(IRanges)
library(rtracklayer) 
```

# Extracting intron annotation from gtf file

```{r}
#####################
## Loading in data ##

# Import GTF file as GRanges
#gtf33 <- rtracklayer::import("/path/reference/hg38/ctat_hg38/GRCh38_gencode_v33_CTAT_lib_Apr062020.plug-n-play/ctat_genome_lib_build_dir/ref_annot.gtf")
gtf33 <- rtracklayer::import("path/hg38/software_indexes/star_fusion/GRCh38_gencode_v33_CTAT_lib_Apr062020.plug-n-play/ctat_genome_lib_build_dir/ref_annot.gtf")
gtf33_df <- as.data.frame(gtf33)

gtf33_trim <- gtf33[which(gtf33$gene_type == "protein_coding")]
gtf33_trim <- gtf33_trim[which(gtf33_trim$type != "Selenocysteine")]
gtf33_trim <- gtf33_trim[which(gtf33_trim$transcript_type == "protein_coding")]
rtracklayer::export(object = gtf33_trim, con = "path/only_PCoding_ctat.gtf", format = "gtf")
gtf33_df <- as.data.frame(gtf33_trim)

# Create txdb object
gtf_ctat_v33 <- file.path("/path/only_PCoding_ctat.gtf")
txdb <- makeTxDbFromGFF(file = gtf_ctat_v33, format = "gtf")

#########################################################################
## Extracting all non-overlapping introns, which doesn't overlap exons ##

# Note: Some regions aren't real introns, but part of intron that doesn't overlap exon in other isoforms.

# Two GRange-sets of non-overlapping introns and exons are created.
# Overlap in ranges between these two sets are found and removed.

txdb_exon <- exonicParts(txdb)
txdb_intron <- intronicParts(txdb)                         # Find all non-overlapping intron regions

hits <- findOverlaps(txdb_intron, txdb_exon)               # Find all exon-intron overlap
intronic_parts3 <- txdb_intron[-queryHits(hits)]           # Remove exon overlapping introns

introns_non_exon_overlap <- as.data.frame(intronic_parts3) # 184,108 introns 
introns_non_exon_overlap$gene_id <- as.character(introns_non_exon_overlap$gene_id)
introns_non_exon_overlap <- left_join(introns_non_exon_overlap, unique(gtf33_df[,-c(2:4,6:9,13:26)]), by=c("seqnames","strand", "gene_id")) # Adding gene names based on matching gene ID

## Make BED file ##
bed_introns <- tibble(chrom = introns_non_exon_overlap$seqnames,
                      chromStart = introns_non_exon_overlap$start,
                      chromEnd = introns_non_exon_overlap$end,
                      name = paste0(introns_non_exon_overlap$seqnames, sep = ":", introns_non_exon_overlap$start, sep = "-",
                                    introns_non_exon_overlap$end, sep = "_", introns_non_exon_overlap$gene_name),
                      score = introns_non_exon_overlap$width,
                      strand = introns_non_exon_overlap$strand)

rtracklayer::export(object = bed_introns, con = "/path/hg38v33_ctat_introns_noOverlap.bed", format = "bed") # Note that 1 is subtracted from start position (converted from 1-based to 0-based).

## Make GTF file ##
introns_non_exon_overlap <- as.data.frame(intronic_parts3) 

introns_non_exon_overlap$gene_id <- as.character(introns_non_exon_overlap$gene_id)
introns_non_exon_overlap2 <- left_join(introns_non_exon_overlap, unique.data.frame(gtf33_df[,-c(2:4,6:9,11,13:26)]), by=c("seqnames","strand", "gene_id")) # Adding gene names
introns_non_exon_overlap2 <- unique.data.frame(introns_non_exon_overlap2)

gtf_introns <- tibble(seqnames = introns_non_exon_overlap2$seqnames,
                      source = "HAVANA", # Not necessarily all found from HAVANA! (just to insert something) 
                      type = "exon", # Note that ALL are introns, we just need to write exon for HTseq to work!
                      start = introns_non_exon_overlap2$start,
                      end = introns_non_exon_overlap2$end,
                      score = ".",
                      strand = introns_non_exon_overlap2$strand,
                      gene_id = introns_non_exon_overlap2$gene_id,
                      gene_type = "protein_coding", # Removed everything that wasn't protein coding
                      gene_name = introns_non_exon_overlap2$gene_name,
                      intron_id = paste0(introns_non_exon_overlap2$seqnames, sep = ":", introns_non_exon_overlap2$start, sep = "-",
                                    introns_non_exon_overlap2$end, sep = "_", introns_non_exon_overlap2$gene_name))

gtf_introns <- gtf_introns[,-c(9,11)] #Remove Gene_id and Gene_name that HTSeq can't process 

rtracklayer::export(object = gtf_introns, con = "/path/hg38v33_ctat_introns_noOverlap.gtf", format = "gtf")
```

Make file to extract phyloP scores from UCSC genome browser:

```{r}
intron_bed <- rtracklayer::import("/path/hg38v33_ctat_introns_noOverlap.bed", format = "bed")

intron_df <- as.data.frame(intron_bed) 
intron_df <- intron_df[, c(1:3,6)]

# Only interested introns from expressed genes.

# Export
rtracklayer::export(object = intron_df, con = "/path/exp_introns.bed", format = "bed")
```


# Extract gene regions and save as .bed file (for Coverage analysis)

Used to extract strand specific regions from .bam file.

```{r}
# Load GTF file
gtf33 <- rtracklayer::import("/path/reference/hg38/ctat_hg38/GRCh38_gencode_v33_CTAT_lib_Apr062020.plug-n-play/ctat_genome_lib_build_dir/ref_annot.gtf")
gtf33_df <- as.data.frame(gtf33)

gene_gtf33_df <- gtf33_df %>% filter(type == "gene")
gene_gtf33_df <- gene_gtf33_df[, c(1:5, 10:12)]
gene_gtf33_df <- gene_gtf33_df %>% filter(gene_type == "protein_coding")

# Extract gene regions

# Create BED format
bed_genes <- tibble(chrom = gene_gtf33_df$seqnames,
                      chromStart = gene_gtf33_df$start,
                      chromEnd = gene_gtf33_df$end,
                      name = paste0(gene_gtf33_df$seqnames, sep = ":", gene_gtf33_df$start, sep = "-",
                                    gene_gtf33_df$end, sep = "_", gene_gtf33_df$gene_name),
                      score = gene_gtf33_df$width,
                      strand = gene_gtf33_df$strand)

# Save BED file
rtracklayer::export(object = bed_genes, con = "/path/hg38v33_ctat_genes_noOverlap.bed", format = "bed") # Note that 1 is subtracted from start position (converted from 1-based to 0-based).
```


# Extract exon regions and save as .bed file (for circular sisRNA junction tracks)

```{r}
# Load GTF file
gtf33 <- rtracklayer::import("/path/hg38/software_indexes/star_fusion/GRCh38_gencode_v33_CTAT_lib_Apr062020.plug-n-play/ctat_genome_lib_build_dir/ref_annot.gtf")
gtf33_df <- as.data.frame(gtf33)

gtf33_df <- gtf33_df %>% filter(type == "exon")
gtf33_df <- gtf33_df[, c(1:5, 10:12)]

# Extract gene regions

# Create BED format
bed_exons <- tibble(chrom = gtf33_df$seqnames,
                      chromStart = gtf33_df$start,
                      chromEnd = gtf33_df$end,
                      name = paste0(gtf33_df$seqnames, sep = ":", gtf33_df$start, sep = "-",
                                    gtf33_df$end, sep = "_", gtf33_df$gene_name),
                      score = gtf33_df$width,
                      strand = gtf33_df$strand)

# Save BED file
rtracklayer::export(object = bed_exons, con = "/path/hg38v33_ctat_exons_noOverlap.bed", format = "bed") # Note that 1 is subtracted from start position (converted from 1-based to 0-based).
```
